### PR TITLE
Issue #4012: add offline-online SAC smoke lane

### PR DESCRIPTION
**Pull request by ll7**
## Reviewer-critical summary

What changed: Adds the first SAC-based offline-to-online RL smoke lane for #4012: `RLTrajectoryDataset.v1` train-split loading, strict observation/action preflight against the online SAC environment, replay-buffer seeding plus optional offline gradient warm-up, online SAC fine-tuning continuation, and a matched from-scratch SAC comparator wrapper.

Why now: the 2026-07-02 ready-queue request asked for the smallest runnable implementation lane, and the earlier issue comment explicitly scoped the first implementation to SAC instead of PPO or Decision Transformer.

Claim boundary: diagnostic implementation smoke only. This PR makes no robustness, benchmark, dissertation, or paper-facing performance claim. Local `output/` datasets/checkpoints remain ignored smoke artifacts, not durable evidence.

Required validation: focused training contract tests plus a tiny CPU smoke of the offline-online and scratch comparator wrapper. Full `pr_ready_check` currently fails on an unrelated baseline lint error in untouched `robot_sf/planner/prediction_mpc.py`.

Remaining blocker: a late maintainer comment at `2026-07-02T15:25:04Z` asks for a broader phased lane with standalone offline pretraining/checkpoint manifests (`pretrain_offline_policy.py`, `offline_to_online_finetune.py`, durable evidence under `docs/context/evidence/...`). This PR does not claim that broader checkpoint-pretraining phase; it is a progression slice that proves SAC replay warm-start plus online fine-tuning wiring.

## Contract delta

- New SAC config block: `offline_online.enabled`, `dataset_path`, `manifest_path`, `dataset_split`, `min_transitions`, `offline_gradient_steps`, `action_contract`, and `observation_contract`.
- New fail-closed blockers: missing enabled dataset path, missing configured manifest path, empty requested split, inconsistent action shapes, non-finite rewards/actions, too few accepted transitions, observation/action mismatch against online env spaces, and invalid orchestrator arm pairing.
- Compatibility impact: existing SAC configs keep `offline_online.enabled == false` by default and preserve existing behavior. New smoke configs are CPU-pinned and diagnostic-only.
- New capability toward #4012: SAC can seed its off-policy replay buffer from validated `RLTrajectoryDataset.v1` transitions, perform optional offline gradient warm-up, then continue online SAC fine-tuning; a paired scratch arm records diagnostic-only comparison metadata.

## Linked Issues

- Relates #4012

## Scope Summary

In scope:
- `robot_sf/training/offline_online_rl.py`: dataset-to-transition loader, provenance preflight, env-space validation, SB3 replay-buffer seeding.
- `robot_sf/training/hybrid_replay_buffer.py`: small deterministic offline/online partition sampler for smoke-level tests and future wiring.
- `scripts/training/train_sac_sb3.py`: opt-in `offline_online` SAC config parsing and pre-learn replay seeding/warm-up.
- `scripts/training/run_offline_online_rl.py`: diagnostic orchestrator for offline-online and scratch SAC arms.
- `configs/training/offline_online_rl/`: CPU smoke configs for the two arms and wrapper.
- `docs/context/issue_4012_offline_online_rl.md`: claim boundary, artifact disposition, and next empirical action.

Out of scope:
- No full benchmark campaign run.
- No Slurm/GPU submission.
- No paper/dissertation claim edits.
- No Decision Transformer implementation.
- No standalone behavior-cloning pretraining checkpoint lane from the late comment.
- No tracked transient queue-routing state.

## Validation / Proof

- PASS: `uv run ruff check robot_sf/training/offline_online_rl.py robot_sf/training/hybrid_replay_buffer.py scripts/training/train_sac_sb3.py scripts/training/run_offline_online_rl.py tests/training/test_offline_online_rl.py tests/training/test_hybrid_replay_buffer.py tests/training/test_train_sac_sb3_offline_online.py tests/training/test_run_offline_online_rl.py`
- PASS: `uv run --extra training pytest tests/training/test_offline_online_rl.py tests/training/test_hybrid_replay_buffer.py tests/training/test_train_sac_sb3_offline_online.py tests/training/test_run_offline_online_rl.py tests/training/test_train_sac_sb3_config.py -q`
- PASS: `git diff --check`
- PASS: `uv run --extra training python scripts/training/run_offline_online_rl.py --config configs/training/offline_online_rl/issue_4012_sac_smoke.yaml`
  - Summary confirmed: `offline_online_rl_comparison.v1`, `diagnostic-smoke-only`, `eligible_for_claim=false`.
- PENDING: refreshed CI after gate formatting fix; prior local focused checks pass. PARTIAL historical: `BASE_REF=origin/main scripts/dev/pr_ready_check.sh` via compact wrapper failed before branch-specific tests on untouched baseline lint:
  - `robot_sf/planner/prediction_mpc.py:253:9: error[F811] Redefinition of unused diagnostics from line 169`
  - Artifact: `.git/codex-agent-runs/active/issue-4012/pr_ready/validation-20260702T152630Z-2060778.summary.json`

## Artifact Disposition

The CPU smoke generated ignored worktree-local files under `output/issue_4012_offline_online_rl_smoke/`, including a tiny dataset, manifest, diagnostic summary/report, and two checkpoints. They were used only to prove the local entry point runs and are not committed or treated as durable evidence.

<details>
<summary>Governance appendix</summary>

Research Result Guidance:
- Target claim / hypothesis / blocker: unblock the implementation path for offline-to-online social-navigation RL by proving SAC can consume validated trajectory rows and continue online fine-tuning.
- Comparator or baseline: paired from-scratch SAC arm with same seed/timesteps/config budget.
- Evidence tier: targeted smoke / diagnostic probe.
- Result classification: diagnostic-only.
- Decision stop rule: do not promote to benchmark or paper-facing claim until durable dataset/checkpoint provenance and a predeclared benchmark campaign exist.
- Parent issue / context note: #4012 and `docs/context/issue_4012_offline_online_rl.md`.
- New research/benchmark/metric/paper-facing analysis tool: no benchmark interpretation tool; support/orchestration only.

Domain-Aware Approval:
- Approval: gate-approved for diagnostic implementation-smoke scope only in the 2026-07-03 batch PR gate; not approval for benchmark, robustness, dissertation, or paper-facing claims.
- Fallback/degraded exclusions: incompatible offline rows fail closed; fallback/degraded rows are not success evidence.
- Claim boundary: diagnostic smoke only.

Falsification / Non-Transfer Check:
- Did mechanism activate? yes, replay buffer seeded from one validated offline transition in CPU smoke.
- Result route: continue with late-comment standalone pretraining/checkpoint manifest work if reviewer wants that broader lane next.

Next Empirical Action:
- Rerun needed: yes, with a durable compatible dataset and explicit benchmark campaign before any performance claim.
- Stop / revise / continue decision: continue only as diagnostic implementation unless the broader pretraining-checkpoint phase is implemented.

Downstream Propagation:
- Parent issue updated: no issue comment by this PR, per user authorization `comment_issue_or_pr: false`; status is included in this PR body.
- Claim map / benchmark report: no, no benchmark claim.
- Leaderboard / artifact catalog: no, local ignored artifacts only.
- Registry or config index: no dedicated index found/needed for this first smoke config.
- Context index / memory note: no index update; added scoped context note.
- Follow-up issue opened: no; deferred standalone checkpoint-pretraining work remains tracked by open parent #4012 and this PR intentionally does not close it.

Risks / Rollout:
- Compatibility risks: low for existing SAC configs because the new block defaults disabled.
- Failure modes: offline dataset shape/space mismatch, missing manifest, too few train transitions, SB3 lifecycle mismatch.
- Rollback fallback plan: revert this PR; existing SAC training path remains unchanged when `offline_online` is absent.

</details>

### CODE REVIEW

**Comment by ll7** [comment]
**Draft-reserved pending correctness fixes (external process review 2026-07-02-v2; platform-enforced per gate policy).** The baseline F811 blocker is resolved on main, but these review-level correctness concerns must be fixed before merge — they are central to replay semantics, not cosmetic:

1. Preserve terminal/truncated transition semantics (final outcomes must not be silently dropped).
2. `HybridReplayBuffer.sample(batch_size)` must return the requested batch size or fail closed (current logic can under-fill with a small offline partition).
3. Preprocess offline observations through the same online-env observation path before space validation.
4. `manifest_path` is REQUIRED for enabled offline-online runs (provenance is not optional here).

Un-draft after a fix commit addresses all four; the hardened gate contract applies (runtime/training semantics).

**Comment by ll7** [comment]
Draft-reserve contract resolved in commit `a3fba18f4` (`fix: satisfy offline-online SAC replay contract`):

1. Terminal/truncated transition semantics -> `a3fba18f4`, `test_load_train_split_derives_next_observations_and_keeps_terminal`, `test_terminal_truncated_offline_transition_survives_hybrid_sample`.
   - Loader keeps final terminal rows, SB3 seeding preserves timeout metadata, and `HybridReplayBuffer.sample()` now returns `truncated` alongside `dones`.

2. `HybridReplayBuffer.sample(batch_size)` exact-size or fail-closed -> `a3fba18f4`, `test_offline_only_sampling_fails_closed_when_partition_too_small`.
   - Offline and online partitions are sampled without replacement and raise `ValueError` when the requested partition count exceeds available rows.

3. Offline observations use online-env preprocessing before validation -> `a3fba18f4`, `test_offline_online_validation_uses_processed_observations`.
   - `_seed_offline_online_replay_buffer()` transforms loaded offline observations via the SAC observation path before `validate_batch_against_env_spaces()`.

4. `manifest_path` required when offline-online mode is enabled -> `a3fba18f4`, `test_sac_config_requires_offline_online_manifest_path`.
   - Config loading now fails closed with `offline_online.manifest_path is required when enabled`.

Validation run:
- `uv run pytest tests/training/test_hybrid_replay_buffer.py tests/training/test_offline_online_rl.py tests/training/test_train_sac_sb3_offline_online.py tests/training/test_run_offline_online_rl.py`
- `uv run pytest tests/training`
- `uv run ruff check robot_sf/training/hybrid_replay_buffer.py scripts/training/train_sac_sb3.py tests/training/test_hybrid_replay_buffer.py tests/training/test_offline_online_rl.py tests/training/test_train_sac_sb3_offline_online.py`
